src/main/java/org/apache/commons/imaging/formats/png/ChunkReader.java

```diff
@@ -0,0 +1,144 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.commons.imaging.formats.png;
+
+import java.io.DataInputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.ArrayList;
+import java.util.List;
+import org.apache.commons.imaging.ImageReadException;
+import static org.apache.commons.imaging.common.BinaryFunctions.readAndVerifyBytes;
+import org.apache.commons.imaging.common.bytesource.ByteSource;
+import org.apache.commons.imaging.formats.png.chunks.ChunkLoader;
+import org.apache.commons.imaging.formats.png.chunks.PngChunk;
+import org.apache.commons.imaging.formats.png.chunks.PngChunkIhdr;
+
+// Replaces readChunk in PngImageParser.
+// This should be a public API. This will allow 3rd party apps
+// to read chunks. Shouldn't ever break.
+
+/**
+ * {@code PngReader} objects can be used to read PNG chunks from
+ * an input stream.
+ * 
+ * @author Shukant Pal
+ */
+public class ChunkReader {
+    
+    DataInputStream input;
+    
+    /**
+     * Asserts that the input stream contains the PNG signature at the
+     * very beginning; throws an ImageReadException otherwise.
+     * 
+     * @throws IOException
+     * @throws ImageReadException 
+     */
+    private void assertSignature() throws IOException, ImageReadException {
+        readAndVerifyBytes(input, PngConstants.PNG_SIGNATURE,
+                "Not a Valid PNG Segment: Incorrect Signature");
+    }
+    
+    public ChunkReader() {    
+    }
+    
+    public ChunkReader(InputStream input) throws ImageReadException, IOException {
+        setInput(input);
+    }
+    
+    public final DataInputStream getInput() {
+        return input;
+    }
+    
+    public final void setInput(InputStream input) throws ImageReadException, IOException {
+        this.input = new DataInputStream(input);
+        assertSignature();
+    }
+    
+    /**
+     * Reads all the chunks (satisfying any filters) from the input
+     * stream containing a PNG file.
+     * 
+     * @param selector can be used to filter chunks
+     * @param returnAfterFirst
+     * @return
+     * @throws IOException
+     * @throws ImageReadException 
+     */
+    public List<PngChunk> readChunks(
+            final ChunkLoader.ChunkSelector selector, final boolean returnAfterFirst)
+            throws IOException, ImageReadException {
+        final List<PngChunk> result = new ArrayList<>();
+        
+        while(true) {
+            final ChunkLoader nextChunkLoaded = ChunkLoader.run(input, selector);
+            PngChunk loadedChunk = nextChunkLoaded.loadedChunk();
+            
+            if(loadedChunk != null) {
+                { // (Shukant Pal): I essentially hate how developers
+                  // have created a dumb PngCrc object. Couldn't it just be
```

Comment:
This wording in a comment is not appropriate for an Open Source project. Shaming any kind of developer and behaviour is not a good idea.